cranelift-wasm/src/code_translator.rs

```diff
@@ -975,9 +975,17 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
                 builder,
             ))
         }
+        Operator::V8x16Shuffle { lanes, .. } => {
+            let (vector_a, vector_b) = state.pop2();
+            let a = optionally_bitcast_vector(vector_a, I8X16, builder);
+            let b = optionally_bitcast_vector(vector_b, I8X16, builder);
+            let mask = builder.func.dfg.immediates.push(lanes.to_vec());
+            let shuffled = builder.ins().shuffle(a, b, mask);
+            state.push1(shuffled)
+            // TODO the original type of a and b are lost here...
```

Comment:
But which type do I use, though: `a` or `b`? The problem is that `v128` in WASM can morph into any type so I chose to `optionally_bitcast_vector` to the type needed by the WASM op (see `type_of(op)` in `ReplaceLane` above, e.g.) before use. This means that an `i32x4.add` could be immediately followed by a `f64x2.replace_lane` and cranelift's implementation would not throw type errors because CLIF would have appropriate bitcasts between. How serious of an issue do you think this is?

Reply:
Good call. Well, the lack of typing in wasm makes it basically impossible to keep a "right" type, because there's no such thing: everything is the wasm SIMD type. I guess a rule of thumb could be to try to avoid creating raw_bitcast instructions until we're proven we don't have any other choices than having them, like here. So maybe we can transform this TODO into a comment saying that we can't infer the output type, and let the wasm->clif converter do the right thing before the next instruction.

(I wonder if this will create type confusion when we have SIMD types in EBB parameters later; probably we could then have the same policy of inserting raw_bitcast before setting EBB parameters.)